### ISSUE TITLE
sails-mongo: Multiple connections defined in connections.js results in incorrect host

### ISSUE THREAD

**Issue by sailsbot** [balderdashy/sails#4019]
_From @abductedMind on August 31, 2015 15:2_

When I have multiple sails-mongo connections defined in connections.js the wrong settings can be used.

In my case I was following the Platzi Sails.js course and hit two road blocks getting the app to run on Heroku. First I had to increase the hookTimeout so that MongoLab would load, then I had to comment out my "localhost" mongodb connection definition. If I did not comment out the unused mongodb connection definition, the wrong host was getting used.

This seems like a bug with either Sails.js or sails-mongo.

_Copied from original issue: balderdashy/sails-mongo#329_

**Comment by sailsbot** [comment]
Thanks for posting, @abductedMind.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
Thanks for posting, @abductedMind.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
_From @constantx on November 28, 2015 5:22_

I'm having this issue as well :(

cc @lebahoang

**Comment by sailsbot** [comment]
_From @particlebanana on January 25, 2016 19:40_

Can someone post their config so I can see whats going on?

**Comment by sailsbot** [comment]
_From @abductedMind on January 25, 2016 20:9_

This file has a commented out connection called "localMongoDb". If I uncomment the definition, no matter what I have set in models.js, "localMongoDb" is used.

[connections.zip](https://github.com/balderdashy/sails-mongo/files/103854/connections.zip)

**Comment by sailsbot** [comment]
_From @particlebanana on January 25, 2016 20:50_

Sorry just to clarify if you have [this config](https://gist.github.com/particlebanana/60fc1d0b8b9c346963ae) you are saying no matter what you put in `config/models.js` for connection it will always use `localMongoDb`?

If so that sounds more like a Sails config issue but I can't recreate it. If that fails will you post you version of Sails, Waterline and Sails-Mongo?

**Comment by sailsbot** [comment]
_From @abductedMind on January 25, 2016 21:21_

Close, but not exactly. This issue was noticed by me months ago so I don't recall and level of testing/experimenting I performed. If in models.js I have the remote mongo source set as being used, it still tries to use the local, if the local is not commented out.

My theory superficially is that when you define multiple mongo connections the last one sticks, even if there are multiple names defined in the connections.js config file.

**Comment by sailsbot** [comment]
_From @wrongey on February 17, 2016 22:41_

I am having the same issue. When the below connections are in my connections.js file, I get the below error when using sails lift. After thinking about it some more, process.env.MONGOLAB_URI is undefined on my development machine. Does sails-mongo try each connection that is defined and fails if they don't work?:

connections.js
`
  developmentCon: {
    adapter: 'sails-mongo',
    host: 'xxxxxx.mongolab.com',
    port: 'xxxxx',
    user: 'xxxx',
    password: 'xxxx',
    database: 'xxx'
  },
  productionCon: {
    adapter: 'sails-mongo',
    url: process.env.MONGOLAB_URI
`

models.js
`
connection: (process.env.MONGOLAB_URI == undefined) ? 'developmentCon' : 'productionCon',
`

`error: A hook (`orm`) failed to load!
error: Error: Failed to connect to MongoDB.  Are you sure your configured Mongo instance is running?
 Error details:
{ [MongoError: connect ECONNREFUSED 127.0.0.1:27017]
  name: 'MongoError',
  message: 'connect ECONNREFUSED 127.0.0.1:27017' }
    at _createError (C:\Projects\SCAFA\node_modules\sails-mongo\lib\adapter.js:103:23)
    at C:\Projects\SCAFA\node_modules\sails-mongo\lib\adapter.js:106:13
    at C:\Projects\SCAFA\node_modules\sails-mongo\lib\connection.js:25:20
    at C:\Projects\SCAFA\node_modules\mongodb\lib\mongo_client.js:276:20
    at C:\Projects\SCAFA\node_modules\mongodb\lib\db.js:224:14
    at null.<anonymous> (C:\Projects\SCAFA\node_modules\mongodb\lib\server.js:234:9)
    at g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at emit (events.js:172:7)
    at null.<anonymous> (C:\Projects\SCAFA\node_modules\mongodb-core\lib\topologies\server.js:269:68)
    at g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at emit (events.js:172:7)
    at null.<anonymous> (C:\Projects\SCAFA\node_modules\mongodb-core\lib\connection\pool.js:81:12)
    at g (events.js:260:16)
    at emitTwo (events.js:87:13)`

**Comment by sailsbot** [comment]
_From @hammady on February 24, 2016 8:8_

Faced same issue.
The problem is that in `lib/connection.js` the `_buildConnection` function will not only construct the connection string, but it will _establish_ a connection as well.

```
MongoClient.connect(connectionString, connectionOptions, cb);
```

It should only construct the connection string, and delay the connection establishment till the correct configuration is selected.

A quick solution is to comment out unwanted connections and leave alone the production config before deploying.

**Comment by sailsbot** [comment]
_From @jsvg on March 6, 2016 23:1_

Facing this issue too. Cannot define multiple sails-mongo connections (config.connections) and assign them separately to config.env.development and config.env.production. Despite modifying configuration in config.models, and elsewhere, only on sails-mongo connection seems achievable. Current workaround, as @hammady mentioned is to just comment disable at deploy.

config.connections is below. regardless of selection (in config.models, or in config.env.dev/prod), sails throws the error @wrongey mentioned.
`
module.exports.connections = {
    mongoDbProf: {
        adapter: "sails-mongo",
        url: process.env.DB_URL
    },
    mongoDbDev: {
        adapter: "sails-mongo",
        host: "localhost",
        port: 1234
        database: "db"
}`

Update: easiest to just define production adapted in config.connections, then in config.local, redefine that adapter for dev db. config.local settings will overwrite config.connections parameters. i.e config.connections is:

`module.exports.connections = { mongo: { adapter: "sails-mongo", url: process.env.DB_URL } }`

config.local is:

`module.export = { connections: { mongo : { adapter: "sails-mongo", etc } } }`

then when deploying, config.local is git ignored by default, so config.connections is exposed.

**Comment by sailsbot** [comment]
_From @bountyx on May 14, 2016 12:58_

+1 Same issue. My hacky work around was a ternary expression dynamically defining the mongo json connection based on the environment. Here's an example of my config/connections.js file:

```
module.exports.connections = {
  mongo: (process.env.NODE_ENV === "production") ?
                  {
                     adapter: 'sails-mongo',
                     host: 'production-server',
                     port: 27017,
                     user: '',
                    password: '',
                    database: 'prod'
                 } : {
                    adapter: 'sails-mongo',
                    host: '127.0.0.1',
                    port: 27017,
                    user: '',
                    password: '',
                   database: 'test'
               }
};
```

**Comment by sgress454** [comment]
From the notes in the [sails.config.connections reference docs](http://sailsjs.com/documentation/reference/configuration/sails-config-connections#?multiple-connections-for-an-adapter):

> **Note** If *any* connection to an adapter is used by a model, then *all* connections to that adapter will be loaded on `sails.lift`, whether or not models are actually using them...It is therefore good practice to split your connection configurations up by environment and save them to the appropriate [environment-specific config files](http://sailsjs.com/documentation/concepts/Configuration?q=environment-specific-files-configenv), or else comment out any connections that you don't want active.

In Sails 1.0, this behavior is clearer: every datastore configuration you define for an environment will be loaded, regardless of whether a model is set up to use it.  This is because in Sails 1.0, you can use a datastore outside of the context of a model.  

Regardless, for Sails v0.12.x, the best practice is to pick a name for a connection (e.g. `mainDb` or `defaultDb`) and then define it in `config/development.js`, `config/production.js` and (optionally) `config/local.js`.  The `config/production.js` file won't be loaded in development, and vice-versa.  Trying to define your development and production connections in the same file will likely lead to the errors reported in this issue!
